### PR TITLE
build: ship esm bundle

### PR DESCRIPTION
**Pull request by DevVersion**
* Now ships a Flat / ESM bundle (similar as for Angular 4+ releases)
* This a big step into cleaning-up the build process. (dynamic extra compiler options)
  This means that we don't need `X` similar tsconfig files for the same module.
* Fixes sourcemap issues with `gulp-autoprefixer` and `gulp-clean-css`.

**Note**: This is not the full-switch to ESM bundles - In future PRs we will remove the single ES-files and can just link to the ES bundle. Similar as in angular.

References #3316

### CODE REVIEW

**Comment by DevVersion** [comment]
I did some refactoring of the gulp tasks. We no longer need multiple `tsconfig` files for the same module.

File: package.json [hunk 1]
```diff
@@ -51,6 +51,7 @@
     "@types/node": "^7.0.4",
     "@types/run-sequence": "^0.0.28",
     "@types/rx": "2.5.33",
+    "autoprefixer": "^6.7.6",
```
Comment (DevVersion) [inline package.json]: Removed all the autoprefixer stuff. So I can just rebase without conflicts if it's merged.

File: tools/gulp/tasks/components.ts [hunk 2]
```diff
@@ -1,139 +1,150 @@
...
   DIST_COMPONENTS_ROOT, PROJECT_ROOT, COMPONENTS_DIR, HTML_MINIFIER_OPTIONS, LICENSE_BANNER
 } from '../constants';
 import {
-  sassBuildTask, tsBuildTask, execNodeTask, copyTask, sequenceTask,
+  sassBuildTask, tsBuildTask, execNodeTask, sequenceTask,
   triggerLivereload
 } from '../task_helpers';
 
-// No typings for these.
+// These imports lack of types.
```
Comment (DevVersion) [inline tools/gulp/tasks/components.ts]: Sounds good

File: tools/gulp/tasks/components.ts [hunk 3]
```diff
@@ -1,139 +1,150 @@
...
-task(':watch:components', () => {
-  watch(path.join(COMPONENTS_DIR, '**/*.ts'), ['build:components', triggerLivereload]);
-  watch(path.join(COMPONENTS_DIR, '**/*.scss'), ['build:components', triggerLivereload]);
-  watch(path.join(COMPONENTS_DIR, '**/*.html'), ['build:components', triggerLivereload]);
-});
-
+/** Builds components for Angular Material releases */
+task(':build:components:release', sequenceTask(
+  ':build:components:bundle:umd',
+  ':build:components:bundle:es',
```
Comment (DevVersion) [inline tools/gulp/tasks/components.ts]: Yeah. Makes sense!

File: tools/gulp/util/ts-compiler.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+/** Formats the TypeScript diagnostics into a error string. */
+export function formatDiagnostics(diagnostics: ts.Diagnostic[]): string {
+  return diagnostics.map(diagnostic => {
+    let res = ts.DiagnosticCategory[diagnostic.category];
+
+    if (diagnostic.file) {
+      let {line, character} = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start);
+
+      res += ' at ' + diagnostic.file.fileName + ':';
+      res += (line + 1) + ':' + (character + 1) + ':';
```
Comment (jelbourn) [inline tools/gulp/util/ts-compiler.ts]: Use template strings?

**Comment by DevVersion** [comment]
@jelbourn  Done! Let me know if you want me to make any other changes.

File: tools/gulp/tasks/components.ts [hunk 5]
```diff
@@ -1,139 +1,150 @@
...
+  ':build:components:bundle:umd',
+  ':build:components:bundle:esm',
+  ':build:components:ngc'
+));
 
-/** Builds component typescript only (ESM output). */
-task(':build:components:ts', tsBuildTask(path.join(COMPONENTS_DIR, 'tsconfig-srcs.json')));
+/** Builds components typescript in ES5, ES6 target. For specs Karma needs CJS output. */
+task(':build:components:ts:es5', tsBuildTask(tsconfigPath, { target: ScriptTarget.ES5 }));
+task(':build:components:ts:es6', tsBuildTask(tsconfigPath, { target: ScriptTarget.ES6 }));
```
Comment (d3viant0ne) [inline tools/gulp/tasks/components.ts]: @DevVersion - Worth noting that `ES6` as a ScriptTarget doesn't work for newer versions of TypeScript iirc.

Options are `ES2015, ES2016, ES2017, ES3, ES5, ESNext & Latest`

Comment (d3viant0ne) [inline tools/gulp/tasks/components.ts]: That that's all of them off the top of my head.

Comment (crisbeto) [inline tools/gulp/tasks/components.ts]: Good point. This does fail on the Angular 4 branch.

Comment (DevVersion) [inline tools/gulp/tasks/components.ts]: Yeah, I noticed it as well, when I based the upcoming packaging PR on top of the v4 compatibility PR. Just FYI: This stuff will be removed with the new version of `tsc-wrapped` anyways.

**Pull request merged**